File: k6/load_tests.js
/**
 * k6/load_tests.js
 * ================
 * Production load test for HOPEFX API.
 *
 * Scenarios
 * ---------
 *   smoke      — 1 VU, 30 s  (CI gate: verify endpoints respond)
 *   load       — ramp 0→50 VUs over 2 min, hold 5 min, ramp down 1 min
 *   soak       — 100 VUs for 30 min  (memory leak / connection pool exhaustion)
 *   spike      — ramp to 500 VUs in 30 s, recover  (thundering herd)
 *   stress     — ramp 0→200 VUs in 5 min, hold 10 min  (find breaking point)
 *   breakpoint — ramp 0→1000 VUs over 20 min  (find absolute limit)
 *
 * Usage
 * -----
 *   k6 run k6/load_tests.js
 *   k6 run --env BASE_URL=https://staging.hopefx.io --env SCENARIO=load k6/load_tests.js
 *   k6 run --env SCENARIO=soak --out json=results/soak.json k6/load_tests.js
 *
 * Environment variables
 * ---------------------
 *   BASE_URL      API base URL (default: http://localhost:8000)
 *   SCENARIO      smoke | load | soak | spike | stress | breakpoint
 *   AUTH_TOKEN    Bearer token for authenticated endpoints
 *   SIGNAL_SYMBOL Symbol for signal/ML tests (default: XAUUSD)
 *   THINK_TIME    Sleep multiplier in seconds (default: 1.0)
 */

import { check, group, sleep } from 'k6';
import http from 'k6/http';
import ws from 'k6/ws';
import { Counter, Rate, Trend } from 'k6/metrics';

// ── Custom metrics ────────────────────────────────────────────────────────────
const errorRate      = new Rate('error_rate');
const orderLatency   = new Trend('order_latency_ms',   true);
const signalLatency  = new Trend('signal_latency_ms',  true);
const mlLatency      = new Trend('ml_latency_ms',      true);
const macroLatency   = new Trend('macro_latency_ms',   true);
const wsConnects     = new Counter('ws_connects');
const wsMessages     = new Counter('ws_messages_received');
const authFailures   = new Counter('auth_failures');
const riskBlocks     = new Counter('risk_blocks');
const ordersFilled   = new Counter('orders_filled');
const rateLimitHits  = new Counter('rate_limit_hits');

// ── Config ────────────────────────────────────────────────────────────────────
const BASE_URL   = __ENV.BASE_URL     || 'http://localhost:8000';
const WS_URL     = BASE_URL.replace(/^http/, 'ws');
const AUTH_TOKEN = __ENV.AUTH_TOKEN   || '';
const SCENARIO   = __ENV.SCENARIO    || 'smoke';
const SIGNAL_SYM = __ENV.SIGNAL_SYMBOL || 'XAUUSD';
const THINK_TIME = parseFloat(__ENV.THINK_TIME || '1.0');
const SYMBOLS    = ['XAUUSD', 'EURUSD', 'GBPUSD', 'USDJPY', 'AUDUSD'];

function randomItem(arr) {
  return arr[Math.floor(Math.random() * arr.length)];
}

// ── Scenario definitions ──────────────────────────────────────────────────────
const SCENARIOS = {
  smoke: {
    executor: 'constant-vus',
    vus: 1,
    duration: '30s',
  },
  load: {
    executor: 'ramping-vus',
    startVUs: 0,
    stages: [
      { duration: '2m',  target: 50  },
      { duration: '5m',  target: 50  },
      { duration: '1m',  target: 0   },
    ],
    gracefulRampDown: '30s',
  },
  soak: {
    executor: 'constant-vus',
    vus: 100,
    duration: '30m',
  },
  spike: {
    executor: 'ramping-vus',
    startVUs: 0,
    stages: [
      { duration: '30s', target: 10  },
      { duration: '30s', target: 500 },
      { duration: '1m',  target: 10  },
      { duration: '30s', target: 0   },
    ],
    gracefulRampDown: '30s',
  },
  stress: {
    executor: 'ramping-vus',
    startVUs: 0,
    stages: [
      { duration: '2m',  target: 50  },
      { duration: '2m',  target: 100 },
      { duration: '2m',  target: 150 },
      { duration: '2m',  target: 200 },
      { duration: '10m', target: 200 },
      { duration: '2m',  target: 0   },
    ],
    gracefulRampDown: '30s',
  },
  breakpoint: {
    executor: 'ramping-arrival-rate',
    startRate: 10,
    timeUnit: '1s',
    preAllocatedVUs: 100,
    maxVUs: 1000,
    stages: [
      { duration: '5m',  target: 50  },
      { duration: '5m',  target: 100 },
      { duration: '5m',  target: 200 },
      { duration: '5m',  target: 500 },
    ],
  },
};

// http_req_failed counts every 4xx as a failure by default, but this suite's own
// checks accept 401/403/404 as valid outcomes (unauthenticated probes, endpoints
// that vary by deployment). Counting them made a <1% threshold unreachable no
// matter how healthy the server was. Narrowing "failed" to 5xx makes the metric
// mean what a load test should measure — the server erroring under load — while
// the per-group checks keep asserting the semantic expectations.
http.setResponseCallback(http.expectedStatuses({ min: 200, max: 499 }));

export const options = {
  scenarios: {
    default: SCENARIOS[SCENARIO] || SCENARIOS.smoke,
  },
  thresholds: {
    http_req_duration:  ['p(95)<500'],   // p95 < 500 ms
    order_latency_ms:   ['p(99)<1000'],  // p99 order < 1 s
    signal_latency_ms:  ['p(95)<300'],   // p95 signal < 300 ms
    ml_latency_ms:      ['p(95)<800'],   // p95 ML inference < 800 ms
    macro_latency_ms:   ['p(95)<400'],   // p95 macro endpoint < 400 ms
    error_rate:         ['rate<0.01'],   // < 1% errors
    http_req_failed:    ['rate<0.01'],
  },
  gracefulStop: '30s',
  noConnectionReuse: false,
  userAgent: 'k6-hopefx-loadtest/2.0',
};

// ── Helpers ───────────────────────────────────────────────────────────────────
function headers(extra) {
  const h = { 'Content-Type': 'application/json', 'Accept': 'application/json' };
  if (AUTH_TOKEN) h['Authorization'] = 'Bearer ' + AUTH_TOKEN;
  return Object.assign(h, extra || {});
}

function checkOk(res, name, allowed) {
  allowed = allowed || [200];
  const ok = check(res, {
    [name + ': status in allowed']: (r) => allowed.indexOf(r.status) !== -1,
    [name + ': not 500']:           (r) => r.status !== 500,
    [name + ': not 502/503/504']:   (r) => [502, 503, 504].indexOf(r.status) === -1,
  });
  errorRate.add(!ok);
  if (res.status === 429) rateLimitHits.add(1);
  return ok;
}

// ── Test functions ────────────────────────────────────────────────────────────

function testHealth() {
  group('health', () => {
    const res = http.get(BASE_URL + '/health', { tags: { name: 'health' }, timeout: '5s' });
    checkOk(res, 'health', [200]);
  });
}

function testPublicStatus() {
  group('public_status', () => {
    const res = http.get(BASE_URL + '/api/status', { tags: { name: 'status' } });
    checkOk(res, 'status', [200, 404]);
  });
}

function testMarketData() {
  group('market_data', () => {
    const symbol = randomItem(SYMBOLS);
    // /api/market-data/<symbol> does not exist and never did — every request
    // to it 404'd, and checkOk accepted 404, so this scenario reported
    // comfortable latency for a route that does no work. The real
    // bid/ask endpoint is /api/trading/prices; OHLCV takes the symbol.
    const res = http.get(BASE_URL + '/api/trading/ohlcv/' + symbol + '?timeframe=M5&limit=100', {
      headers: headers(), tags: { name: 'market_data' },
    });
    checkOk(res, 'market_data', [200, 401, 403, 503]);
  });
}

function testSignalEndpoint() {
  group('signal', () => {
    const start = Date.now();
    const res = http.get(BASE_URL + '/api/signals/latest?symbol=' + SIGNAL_SYM, {
      headers: headers(), tags: { name: 'signal_latest' },
    });
    signalLatency.add(Date.now() - start);
    checkOk(res, 'signal', [200, 401, 403, 404, 503]);
  });
}

function testMLPredict() {
  group('ml_predict', () => {
    const start = Date.now();
    // POST, not GET: /api/ml/predict/{symbol} is registered for POST only, so
    // the GET this used to send never reached the handler and the ML latency
    // budget measured a routing miss.
    const res = http.post(BASE_URL + '/api/ml/predict/' + SIGNAL_SYM, '{}', {
      headers: headers(), tags: { name: 'ml_predict' },
    });
    mlLatency.add(Date.now() - start);
    checkOk(res, 'ml_predict', [200, 401, 403, 404, 503]);
  });
}

function testMLStatus() {
  group('ml_status', () => {
    // /api/ml/status is not a registered route — the app serves
    // /api/ml/drift/status, /api/ml/anomaly/status and /api/ml/rl/status, plus
    // /api/superadmin/ml/status behind the superadmin gate. This probed a 404
    // on every iteration and counted it as a pass.
    const res = http.get(BASE_URL + '/api/ml/drift/status', {
      headers: headers(), tags: { name: 'ml_status' },
    });
    checkOk(res, 'ml_status', [200, 503]);
  });
}

function testMacroSnapshot() {
  group('macro_snapshot', () => {
    const start = Date.now();
    const res = http.get(BASE_URL + '/api/macro/snapshot', {
      headers: headers(), tags: { name: 'macro_snapshot' },
    });
    macroLatency.add(Date.now() - start);
    checkOk(res, 'macro_snapshot', [200, 503]);
  });
}

function testMacroStore() {
  group('macro_store', () => {
    const start = Date.now();
    const res = http.get(BASE_URL + '/api/macro/store', {
      headers: headers(), tags: { name: 'macro_store' },
    });
    macroLatency.add(Date.now() - start);
    checkOk(res, 'macro_store', [200, 503]);
  });
}

function testMacroFeatures() {
  group('macro_features', () => {
    const res = http.get(BASE_URL + '/api/macro/features', {
      headers: headers(), tags: { name: 'macro_features' },
    });
    checkOk(res, 'macro_features', [200, 503]);
  });
}

function testAuthLogin() {
  group('auth_login', () => {
    const res = http.post(
      BASE_URL + '/auth/login',
      JSON.stringify({ email: 'loadtest@example.com', password: 'LoadTest123!' }),  // pragma: allowlist secret
      { headers: headers(), tags: { name: 'auth_login' } },
    );
    const ok = check(res, {
      'login: responds (not 500)': (r) => r.status !== 500,
      'login: responds fast':      (r) => r.timings.duration < 3000,
    });
    if (res.status === 401) authFailures.add(1);
    errorRate.add(!ok);
  });
}

function testOrderEndpoint() {
  if (!AUTH_TOKEN) return;
  group('place_order', () => {
    const symbol = randomItem(SYMBOLS);
    const side   = randomItem(['buy', 'sell']);
    const start  = Date.now();
    const res = http.post(
      BASE_URL + '/api/trading/order',
      JSON.stringify({ symbol, side, quantity: 0.01, order_type: 'market' }),
      { headers: headers(), tags: { name: 'place_order' } },
    );
    orderLatency.add(Date.now() - start);
    const ok = check(res, {
      'order: not 500':         (r) => r.status !== 500,
      'order: expected status': (r) => [201, 400, 401, 403, 422, 429, 503].indexOf(r.status) !== -1,
    });
    errorRate.add(!ok);
    if (res.status === 201) ordersFilled.add(1);
    if (res.status === 403) riskBlocks.add(1);
    if (res.status === 429) rateLimitHits.add(1);
  });
}

function testPositions() {
  if (!AUTH_TOKEN) return;
  group('positions', () => {
    const res = http.get(BASE_URL + '/api/trading/positions', {
      headers: headers(), tags: { name: 'positions' },
    });
    checkOk(res, 'positions', [200, 401, 403, 503]);
  });
}

function testAccountInfo() {
  if (!AUTH_TOKEN) return;
  group('account_info', () => {
    const res = http.get(BASE_URL + '/api/trading/account', {
      headers: headers(), tags: { name: 'account_info' },
    });
    checkOk(res, 'account_info', [200, 401, 403, 404, 503]);
  });
}

function testRiskStatus() {
  group('risk_status', () => {
    // Was /api/risk/status, which is not a registered route. The /api/risk
    // prefix belongs to the prop-firm and calculator routers; the risk metrics
    // snapshot is /api/trading/risk.
    const res = http.get(BASE_URL + '/api/trading/risk', {
      headers: headers(), tags: { name: 'risk_status' },
    });
    checkOk(res, 'risk_status', [200, 401, 403, 503]);
  });
}

function testPrometheusMetrics() {
  group('prometheus', () => {
    const res = http.get(BASE_URL + '/metrics', { tags: { name: 'prometheus_metrics' } });
    checkOk(res, 'prometheus', [200, 403, 404]);
  });
}

/**
 * Rate-limit probe: send 15 rapid requests to an auth endpoint and verify
 * that at least one 429 is returned (rate limiter is active).
 * Only runs in smoke/load scenarios to avoid inflating error counts.
 */
function testRateLimitEnforced() {
  if (SCENARIO !== 'smoke' && SCENARIO !== 'load') return;
  group('rate_limit_probe', () => {
    let got429 = false;
    for (let i = 0; i < 15; i++) {
      const res = http.post(
        BASE_URL + '/auth/login',
        JSON.stringify({ email: 'ratelimit@example.com', password: 'wrong' }),  // pragma: allowlist secret
        {
          headers: headers(),
          tags: { name: 'rate_limit_probe' },
          // This group sends deliberately-wrong credentials and wants a 429.
          // Its rejections are counted by the module-level responseCallback
          // above, which treats every 4xx as expected; enumerating statuses
          // here instead is how this got it wrong first time (the app answers
          // 403, not 401, so all 60 probe requests were counted as failures).
        },
      );
      if (res.status === 429) { got429 = true; break; }
    }
    check({ got429 }, {
      'rate limiter active (got 429 after rapid requests)': (d) => d.got429,
    });
  });
}

/**
 * WebSocket smoke test: connect to /ws/live, wait for one message, disconnect.
 * Counts successful connections and messages received.
 */
function testWebSocket() {
  if (SCENARIO !== 'smoke') return;  // WS test only in smoke to avoid VU exhaustion
  group('websocket', () => {
    const url = WS_URL + '/ws/live';
    const params = AUTH_TOKEN ? { headers: { Authorization: 'Bearer ' + AUTH_TOKEN } } : {};
    const res = ws.connect(url, params, (socket) => {
      wsConnects.add(1);
      socket.on('open', () => {
        socket.send(JSON.stringify({ type: 'subscribe', symbol: SIGNAL_SYM }));
      });
      socket.on('message', (msg) => {
        wsMessages.add(1);
        socket.close();
      });
      socket.on('error', () => { socket.close(); });
      socket.setTimeout(() => { socket.close(); }, 5000);
    });
    check(res, {
      'ws: connected (101 or 200)': (r) => r && (r.status === 101 || r.status === 200),
    });
  });
}

// ── Main VU function ──────────────────────────────────────────────────────────
export default function () {
  testHealth();              sleep(0.1 * THINK_TIME);
  testPublicStatus();        sleep(0.1 * THINK_TIME);
  testMarketData();          sleep(0.2 * THINK_TIME);
  testSignalEndpoint();      sleep(0.2 * THINK_TIME);
  testMLStatus();            sleep(0.1 * THINK_TIME);
  testMLPredict();           sleep(0.3 * THINK_TIME);
  testMacroSnapshot();       sleep(0.1 * THINK_TIME);
  testMacroStore();          sleep(0.1 * THINK_TIME);
  testMacroFeatures();       sleep(0.1 * THINK_TIME);
  testRiskStatus();          sleep(0.1 * THINK_TIME);
  testPrometheusMetrics();   sleep(0.1 * THINK_TIME);
  testAuthLogin();           sleep(0.3 * THINK_TIME);
  testWebSocket();           sleep(0.2 * THINK_TIME);
  testOrderEndpoint();       sleep(0.2 * THINK_TIME);
  testPositions();           sleep(0.2 * THINK_TIME);
  testAccountInfo();         sleep(0.2 * THINK_TIME);
  // Last, deliberately. The limiter is per caller, not per endpoint, so once
  // this group trips it every request after it in the same iteration is 429'd.
  // Running it mid-iteration made all nine later groups fail 5 of 9 checks and
  // read as endpoint outages rather than as the probe doing its job.
  testRateLimitEnforced();   sleep(0.5 * THINK_TIME);
}

// ── Setup ─────────────────────────────────────────────────────────────────────
export function setup() {
  const res = http.get(BASE_URL + '/health', { timeout: '10s' });
  if (res.status !== 200) {
    console.warn('WARNING: /health returned ' + res.status + ' — server may not be ready');
  }
  console.log('Load test starting: scenario=' + SCENARIO + ' base_url=' + BASE_URL);
  if (!AUTH_TOKEN) {
    console.log('INFO: AUTH_TOKEN not set — authenticated endpoints will be skipped');
  }
  return { base_url: BASE_URL, scenario: SCENARIO, started_at: new Date().toISOString() };
}

// ── Teardown ──────────────────────────────────────────────────────────────────
export function teardown(data) {
  console.log(
    'Load test complete: scenario=' + data.scenario +
    ' base_url=' + data.base_url +
    ' started_at=' + data.started_at,
  );
}

File: locust/load_tests.py
# HOPEFX-AI-TRADING
# Copyright (c) 2025-2026
# Licensed under GNU Affero General Public License v3.0 (AGPL-3.0)
# All modifications must be shared under the same license.
# No commercial use without explicit permission.
"""
locust/load_tests.py
====================
Production load test for HOPEFX API using Locust.

User classes
------------
  PublicUser          — unauthenticated visitor (weight 3)
  MLResearcher        — user polling ML/signal endpoints (weight 2)
  AuthenticatedTrader — authenticated trader placing orders (weight 1)

Usage
-----
  # Headless CI run
  locust -f locust/load_tests.py \
         --headless --users 100 --spawn-rate 10 \
         --run-time 5m \
         --host https://api.hopefx.io

  # Interactive web UI (local)
  locust -f locust/load_tests.py --host http://localhost:8000
  # then open http://localhost:8089

  # CSV output for CI artefacts
  locust -f locust/load_tests.py \
         --headless --users 50 --spawn-rate 5 --run-time 2m \
         --csv results/locust \
         --host http://localhost:8000

  # HTML report
  locust -f locust/load_tests.py \
         --headless --users 50 --spawn-rate 5 --run-time 2m \
         --html results/locust_report.html \
         --host http://localhost:8000

Environment variables
---------------------
  AUTH_TOKEN    Bearer token for authenticated endpoints (optional).
  LOCUST_HOST   Overrides --host when set.
  THINK_TIME    Float multiplier for wait_time (default: 1.0).
  SIGNAL_SYMBOL Symbol for ML/signal tests (default: XAUUSD).
"""

from __future__ import annotations

import os
import random

from locust import HttpUser, between, events, task
from locust.exception import StopUser

_AUTH_TOKEN = os.getenv("AUTH_TOKEN", "")
_THINK_TIME = float(os.getenv("THINK_TIME", "1.0"))
_SYMBOLS = ["XAUUSD", "EURUSD", "GBPUSD", "USDJPY", "AUDUSD"]
_SIGNAL_SYM = os.getenv("SIGNAL_SYMBOL", "XAUUSD")


def _auth_headers() -> dict[str, str]:
    h = {"Content-Type": "application/json", "Accept": "application/json"}
    if _AUTH_TOKEN:
        h["Authorization"] = f"Bearer {_AUTH_TOKEN}"
    return h


def _check(resp, name: str, allowed=(200,)) -> bool:
    """Mark response success/failure. Returns True if OK."""
    if resp.status_code in allowed:
        resp.success()
        return True
    if resp.status_code in (500, 502, 503, 504):
        resp.failure(f"{name}: server error {resp.status_code}")
        return False
    resp.success()
    return True


# ── Public user ───────────────────────────────────────────────────────────────


class PublicUser(HttpUser):
    """Unauthenticated visitor hitting public endpoints."""

    wait_time = between(0.5 * _THINK_TIME, 2.0 * _THINK_TIME)
    weight = 3

    @task(5)
    def health_check(self):
        with self.client.get("/health", catch_response=True, name="/health") as r:
            _check(r, "health", (200,))

    @task(3)
    def public_status(self):
        with self.client.get("/api/status", catch_response=True, name="/api/status") as r:
            _check(r, "status", (200, 404))

    @task(4)
    def market_data(self):
        symbol = random.choice(_SYMBOLS)  # nosec B311 - load test symbol selection, not cryptographic
        with self.client.get(
            f"/api/market-data/{symbol}",
            catch_response=True,
            name="/api/market-data/[symbol]",
        ) as r:
            _check(r, "market_data", (200, 404, 503))

    @task(2)
    def prometheus_metrics(self):
        with self.client.get("/metrics", catch_response=True, name="/metrics") as r:
            _check(r, "prometheus", (200, 403, 404))

    @task(2)
    def risk_status(self):
        with self.client.get(
            "/api/risk/status",
            headers=_auth_headers(),
            catch_response=True,
            name="/api/risk/status",
        ) as r:
            _check(r, "risk_status", (200, 401, 403, 404, 503))

    @task(1)
    def auth_login_attempt(self):
        """Test login endpoint responds — 401 expected for non-existent user."""
        with self.client.post(
            "/auth/login",
            json={"email": "loadtest@example.com", "password": "LoadTest123!"},  # nosec B105 - load-test probe, 401 expected
            catch_response=True,
            name="/auth/login",
        ) as r:
            _check(r, "auth_login", (200, 201, 401, 422))


# ── ML researcher ─────────────────────────────────────────────────────────────


class MLResearcher(HttpUser):
    """User polling ML predictions and signal endpoints."""

    wait_time = between(1.0 * _THINK_TIME, 4.0 * _THINK_TIME)
    weight = 2

    @task(4)
    def ml_status(self):
        with self.client.get(
            "/api/ml/status",
            headers=_auth_headers(),
            catch_response=True,
            name="/api/ml/status",
        ) as r:
            _check(r, "ml_status", (200, 401, 403, 404, 503))

    @task(4)
    def ml_predict(self):
        with self.client.get(
            f"/api/ml/predict/{_SIGNAL_SYM}",
            headers=_auth_headers(),
            catch_response=True,
            name="/api/ml/predict/[symbol]",
        ) as r:
            _check(r, "ml_predict", (200, 401, 403, 404, 503))

    @task(3)
    def ml_accuracy(self):
        with self.client.get(
            "/api/ml/accuracy",
            headers=_auth_headers(),
            catch_response=True,
            name="/api/ml/accuracy",
        ) as r:
            _check(r, "ml_accuracy", (200, 401, 403, 404, 503))

    @task(3)
    def signal_latest(self):
        with self.client.get(
            f"/api/signals/latest?symbol={_SIGNAL_SYM}",
            headers=_auth_headers(),
            catch_response=True,
            name="/api/signals/latest",
        ) as r:
            _check(r, "signal_latest", (200, 401, 403, 404, 503))

    @task(2)
    def ml_models_list(self):
        with self.client.get(
            "/api/ml/models",
            headers=_auth_headers(),
            catch_response=True,
            name="/api/ml/models",
        ) as r:
            _check(r, "ml_models", (200, 401, 403, 404, 503))

    @task(1)
    def feature_groups(self):
        with self.client.get(
            "/api/ml/features/groups",
            headers=_auth_headers(),
            catch_response=True,
            name="/api/ml/features/groups",
        ) as r:
            _check(r, "feature_groups", (200, 401, 403, 404, 503))


# ── Authenticated trader ──────────────────────────────────────────────────────


class AuthenticatedTrader(HttpUser):
    """Authenticated trader placing orders and checking positions."""

    wait_time = between(1.0 * _THINK_TIME, 3.0 * _THINK_TIME)
    weight = 1

    def on_start(self):
        if not _AUTH_TOKEN:
            raise StopUser()

    @task(5)
    def get_positions(self):
        with self.client.get(
            "/api/trading/positions",
            headers=_auth_headers(),
            catch_response=True,
            name="/api/trading/positions",
        ) as r:
            _check(r, "positions", (200, 401, 403, 503))

    @task(3)
    def get_account_info(self):
        with self.client.get(
            "/api/trading/account",
            headers=_auth_headers(),
            catch_response=True,
            name="/api/trading/account",
        ) as r:
            _check(r, "account_info", (200, 401, 403, 404, 503))

    @task(2)
    def place_order(self):
        symbol = random.choice(_SYMBOLS)  # nosec B311 - load test symbol selection, not cryptographic
        side = random.choice(["buy", "sell"])  # nosec B311 - load test side selection, not cryptographic
        qty = round(random.uniform(0.01, 0.1), 2)  # nosec B311 - load test quantity, not cryptographic

        with self.client.post(
            "/api/trading/order",
            headers=_auth_headers(),
            json={
                "symbol": symbol,
                "side": side,
                "quantity": qty,
                "order_type": "market",
            },
            catch_response=True,
            name="/api/trading/order",
        ) as r:
            if r.status_code in (201, 400, 403, 422, 429, 503):
                r.success()
            elif r.status_code == 401:  # noqa: PLR2004
                r.failure("Auth token rejected")
            elif r.status_code == 500:  # noqa: PLR2004
                r.failure(f"Server error placing order: {r.text[:200]}")
            else:
                r.success()

    @task(2)
    def get_performance_summary(self):
        with self.client.get(
            "/api/performance/summary",
            headers=_auth_headers(),
            catch_response=True,
            name="/api/performance/summary",
        ) as r:
            _check(r, "performance_summary", (200, 401, 403, 404, 503))

    @task(1)
    def get_risk_metrics(self):
        with self.client.get(
            "/api/risk/metrics",
            headers=_auth_headers(),
            catch_response=True,
            name="/api/risk/metrics",
        ) as r:
            _check(r, "risk_metrics", (200, 401, 403, 404, 503))

    @task(1)
    def get_trade_history(self):
        with self.client.get(
            "/api/trading/history?limit=20",
            headers=_auth_headers(),
            catch_response=True,
            name="/api/trading/history",
        ) as r:
            _check(r, "trade_history", (200, 401, 403, 404, 503))


# ── Event hooks ───────────────────────────────────────────────────────────────


@events.test_start.add_listener
def on_test_start(environment, **kwargs):
    print(f"[locust] Starting load test against {environment.host}")
    if not _AUTH_TOKEN:
        print("[locust] AUTH_TOKEN not set — AuthenticatedTrader users will be skipped")
    print(f"[locust] THINK_TIME multiplier: {_THINK_TIME}x")


@events.test_stop.add_listener
def on_test_stop(environment, **kwargs):
    stats = environment.stats.total
    p95 = stats.get_response_time_percentile(0.95) or 0
    p99 = stats.get_response_time_percentile(0.99) or 0
    failure_rate = stats.num_failures / stats.num_requests * 100 if stats.num_requests > 0 else 0
    print(
        f"[locust] Test complete — "
        f"requests={stats.num_requests} "
        f"failures={stats.num_failures} ({failure_rate:.1f}%) "
        f"p95={p95:.0f}ms "
        f"p99={p99:.0f}ms "
        f"rps={stats.current_rps:.1f}"
    )
    if failure_rate > 1.0:
        print(f"[locust] FAIL: error rate {failure_rate:.1f}% exceeds 1% threshold")
